src/compiler/core.ts

```diff
@@ -409,6 +409,19 @@ namespace ts {
         return hasOwnProperty.call(map, key) ? map[key] : undefined;
     }
 
+    /** Gets a property in a Map, or if it doesn't exist, creates and returns it. */
+    export function getOrUpdate<T>(map: Map<T>, key: string, makeValue: () => T) {
+        return map[key] || (map[key] = makeValue());

```

Comment:
Consider adding `key` to `makeValue`, that way we can use it as a factory to avoid introducing closures.